File: surface/widgets/compose/components/widgets/concepts/llm-core/GenerationIndicator.kt
package com.clef.surface.widgets.concepts.llmcore

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --- State machine ---

enum class GenerationIndicatorState { Idle, Generating, Complete, Error }

sealed class GenerationIndicatorEvent {
    object Start : GenerationIndicatorEvent()
    object Token : GenerationIndicatorEvent()
    object Complete : GenerationIndicatorEvent()
    object Error : GenerationIndicatorEvent()
    object Reset : GenerationIndicatorEvent()
    object Retry : GenerationIndicatorEvent()
}

fun generationIndicatorReduce(state: GenerationIndicatorState, event: GenerationIndicatorEvent): GenerationIndicatorState = when (state) {
    GenerationIndicatorState.Idle -> when (event) {
        is GenerationIndicatorEvent.Start -> GenerationIndicatorState.Generating
        else -> state
    }
    GenerationIndicatorState.Generating -> when (event) {
        is GenerationIndicatorEvent.Token -> GenerationIndicatorState.Generating
        is GenerationIndicatorEvent.Complete -> GenerationIndicatorState.Complete
        is GenerationIndicatorEvent.Error -> GenerationIndicatorState.Error
        else -> state
    }
    GenerationIndicatorState.Complete -> when (event) {
        is GenerationIndicatorEvent.Reset -> GenerationIndicatorState.Idle
        is GenerationIndicatorEvent.Start -> GenerationIndicatorState.Generating
        else -> state
    }
    GenerationIndicatorState.Error -> when (event) {
        is GenerationIndicatorEvent.Reset -> GenerationIndicatorState.Idle
        is GenerationIndicatorEvent.Retry -> GenerationIndicatorState.Generating
        else -> state
    }
}

// --- Helpers ---

private fun formatElapsed(seconds: Int): String {
    if (seconds < 60) return "${seconds}s"
    val m = seconds / 60
    val s = seconds % 60
    return if (s > 0) "${m}m ${s}s" else "${m}m"
}

@Composable
fun GenerationIndicator(
    status: GenerationIndicatorState,
    modifier: Modifier = Modifier,
    model: String? = null,
    tokenCount: Int? = null,
    showTokens: Boolean = true,
    showModel: Boolean = true,
    showElapsed: Boolean = true,
    variant: String = "dots",
    onRetry: () -> Unit = {}
) {
    var state by remember { mutableStateOf(GenerationIndicatorState.Idle) }
    var elapsedSeconds by remember { mutableIntStateOf(0) }
    var finalElapsed by remember { mutableIntStateOf(0) }

    // Sync reducer state with status prop
    LaunchedEffect(status) {
        when (status) {
            GenerationIndicatorState.Generating -> {
                if (state == GenerationIndicatorState.Idle || state == GenerationIndicatorState.Complete || state == GenerationIndicatorState.Error) {
                    state = if (state == GenerationIndicatorState.Error) {
                        generationIndicatorReduce(state, GenerationIndicatorEvent.Retry)
                    } else {
                        generationIndicatorReduce(state, GenerationIndicatorEvent.Start)
                    }
                }
            }
            GenerationIndicatorState.Complete -> {
                if (state == GenerationIndicatorState.Generating) {
                    state = generationIndicatorReduce(state, GenerationIndicatorEvent.Complete)
                }
            }
            GenerationIndicatorState.Error -> {
                if (state == GenerationIndicatorState.Generating) {
                    state = generationIndicatorReduce(state, GenerationIndicatorEvent.Error)
                }
            }
            GenerationIndicatorState.Idle -> {
                if (state == GenerationIndicatorState.Complete || state == GenerationIndicatorState.Error) {
                    state = generationIndicatorReduce(state, GenerationIndicatorEvent.Reset)
                }
            }
        }
    }

    // Elapsed time timer
    LaunchedEffect(state) {
        when (state) {
            GenerationIndicatorState.Generating -> {
                elapsedSeconds = 0
                while (true) {
                    delay(1000)
                    elapsedSeconds++
                }
            }
            GenerationIndicatorState.Complete, GenerationIndicatorState.Error -> {
                finalElapsed = elapsedSeconds
            }
            GenerationIndicatorState.Idle -> {
                elapsedSeconds = 0
                finalElapsed = 0
            }
        }
    }

    val isGenerating = state == GenerationIndicatorState.Generating

    val statusText = when (state) {
        GenerationIndicatorState.Generating -> "Generating\u2026"
        GenerationIndicatorState.Complete -> "Complete"
        GenerationIndicatorState.Error -> "Error"
        else -> ""
    }

    val elapsedText = when (state) {
        GenerationIndicatorState.Generating -> formatElapsed(elapsedSeconds)
        GenerationIndicatorState.Complete, GenerationIndicatorState.Error -> formatElapsed(finalElapsed)
        else -> ""
    }

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier.semantics { contentDescription = "Generation ${state.name.lowercase()}" }
    ) {
        // Spinner
        if (isGenerating) {
            when (variant) {
                "spinner" -> CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
                "bar" -> LinearProgressIndicator(modifier = Modifier.width(48.dp).height(4.dp))
                else -> Text("\u2026", fontSize = 14.sp) // dots
            }
        }

        // Status text
        if (statusText.isNotEmpty()) {
            Text(statusText, fontSize = 13.sp)
        }

        // Model badge
        if (showModel && model != null) {
            Text(model, fontSize = 12.sp, color = MaterialTheme.colorScheme.primary)
        }

        // Token counter
        if (showTokens && tokenCount != null) {
            Text("$tokenCount tokens", fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }

        // Elapsed time
        if (showElapsed && (state == GenerationIndicatorState.Generating || state == GenerationIndicatorState.Complete)) {
            Text(elapsedText, fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }

        // Retry button
        if (state == GenerationIndicatorState.Error) {
            TextButton(onClick = {
                state = generationIndicatorReduce(state, GenerationIndicatorEvent.Retry)
                onRetry()
            }) {
                Text("Retry", fontSize = 12.sp)
            }
        }
    }
}

File: surface/widgets/nativescript/components/widgets/concepts/llm-core/GenerationIndicator.ts
import { StackLayout, Label, Button, FlexboxLayout, ActivityIndicator } from '@nativescript/core';

/* ---------------------------------------------------------------------------
 * GenerationIndicator state machine
 * States: idle, generating, complete, error
 * ------------------------------------------------------------------------- */

export type GenerationIndicatorState = 'idle' | 'generating' | 'complete' | 'error';
export type GenerationIndicatorEvent =
  | { type: 'START' }
  | { type: 'TOKEN' }
  | { type: 'COMPLETE' }
  | { type: 'ERROR' }
  | { type: 'RESET' }
  | { type: 'RETRY' };

export function generationIndicatorReducer(state: GenerationIndicatorState, event: GenerationIndicatorEvent): GenerationIndicatorState {
  switch (state) {
    case 'idle':
      if (event.type === 'START') return 'generating';
      return state;
    case 'generating':
      if (event.type === 'TOKEN') return 'generating';
      if (event.type === 'COMPLETE') return 'complete';
      if (event.type === 'ERROR') return 'error';
      return state;
    case 'complete':
      if (event.type === 'RESET') return 'idle';
      if (event.type === 'START') return 'generating';
      return state;
    case 'error':
      if (event.type === 'RESET') return 'idle';
      if (event.type === 'RETRY') return 'generating';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

function formatElapsed(seconds: number): string {
  if (seconds < 60) return `${seconds}s`;
  const m = Math.floor(seconds / 60);
  const s = seconds % 60;
  return s > 0 ? `${m}m ${s}s` : `${m}m`;
}

/* ---------------------------------------------------------------------------
 * Props
 * ------------------------------------------------------------------------- */

export interface GenerationIndicatorProps {
  status: GenerationIndicatorState;
  model?: string;
  tokenCount?: number;
  showTokens?: boolean;
  showModel?: boolean;
  showElapsed?: boolean;
  variant?: 'dots' | 'spinner' | 'bar';
  cancelable?: boolean;
  onCancel?: () => void;
  onRetry?: () => void;
}

/* ---------------------------------------------------------------------------
 * Widget
 * ------------------------------------------------------------------------- */

export function createGenerationIndicator(props: GenerationIndicatorProps): { view: StackLayout; dispose: () => void } {
  const {
    status,
    model,
    tokenCount,
    showTokens = true,
    showModel = true,
    showElapsed = true,
    cancelable = false,
    onCancel,
    onRetry,
  } = props;

  let state: GenerationIndicatorState = 'idle';
  let elapsedSeconds = 0;
  let finalElapsed = 0;
  let intervalId: ReturnType<typeof setInterval> | null = null;
  const disposers: (() => void)[] = [];

  function send(event: GenerationIndicatorEvent) {
    state = generationIndicatorReducer(state, event);
    update();
  }

  // Sync with status prop
  function syncStatus() {
    switch (status) {
      case 'generating':
        if (state === 'idle' || state === 'complete' || state === 'error') {
          send({ type: state === 'error' ? 'RETRY' : 'START' });
        }
        break;
      case 'complete':
        if (state === 'generating') send({ type: 'COMPLETE' });
        break;
      case 'error':
        if (state === 'generating') send({ type: 'ERROR' });
        break;
      case 'idle':
        if (state === 'complete' || state === 'error') send({ type: 'RESET' });
        break;
    }
  }

  const root = new StackLayout();
  root.className = 'generation-indicator';
  root.automationText = `Generation ${status}`;

  const mainRow = new FlexboxLayout();
  mainRow.className = 'generation-indicator-row';
  mainRow.flexDirection = 'row' as any;
  mainRow.alignItems = 'center' as any;

  // Spinner
  const spinner = new ActivityIndicator();
  spinner.className = 'generation-indicator-spinner';
  spinner.busy = false;
  spinner.width = 20;
  spinner.height = 20;
  spinner.visibility = 'collapse' as any;
  mainRow.addChild(spinner);

  // Status text
  const statusText = new Label();
  statusText.className = 'generation-indicator-status';
  statusText.text = '';
  mainRow.addChild(statusText);

  // Model badge
  const modelBadge = new Label();
  modelBadge.className = 'generation-indicator-model';
  modelBadge.visibility = 'collapse' as any;
  if (showModel && model) {
    modelBadge.text = model;
    modelBadge.visibility = 'visible' as any;
  }
  mainRow.addChild(modelBadge);

  // Token counter
  const tokenLabel = new Label();
  tokenLabel.className = 'generation-indicator-tokens';
  tokenLabel.visibility = 'collapse' as any;
  if (showTokens && tokenCount != null) {
    tokenLabel.text = `${tokenCount} tokens`;
    tokenLabel.visibility = 'visible' as any;
    tokenLabel.automationText = `${tokenCount} tokens generated`;
  }
  mainRow.addChild(tokenLabel);

  // Elapsed time
  const elapsedLabel = new Label();
  elapsedLabel.className = 'generation-indicator-elapsed';
  elapsedLabel.visibility = 'collapse' as any;
  mainRow.addChild(elapsedLabel);

  root.addChild(mainRow);

  // Retry button (error state)
  const retryBtn = new Button();
  retryBtn.className = 'generation-indicator-retry';
  retryBtn.text = 'Retry';
  retryBtn.automationText = 'Retry generation';
  retryBtn.visibility = 'collapse' as any;
  const retryHandler = () => {
    if (state === 'error') onRetry?.();
  };
  retryBtn.on('tap', retryHandler);
  disposers.push(() => retryBtn.off('tap', retryHandler));
  root.addChild(retryBtn);

  function stopInterval() {
    if (intervalId !== null) {
      clearInterval(intervalId);
      intervalId = null;
    }
  }

  function startInterval() {
    stopInterval();
    elapsedSeconds = 0;
    intervalId = setInterval(() => {
      elapsedSeconds += 1;
      if (showElapsed) {
        elapsedLabel.text = formatElapsed(elapsedSeconds);
      }
    }, 1000);
  }

  function update() {
    const isGenerating = state === 'generating';

    // Spinner
    spinner.busy = isGenerating;
    spinner.visibility = (isGenerating ? 'visible' : 'collapse') as any;

    // Status text
    switch (state) {
      case 'generating': statusText.text = 'Generating...'; break;
      case 'complete': statusText.text = 'Complete'; break;
      case 'error': statusText.text = 'Error'; break;
      default: statusText.text = ''; break;
    }

    // Elapsed
    if (showElapsed && (state === 'generating' || state === 'complete')) {
      elapsedLabel.visibility = 'visible' as any;
      if (state === 'generating') {
        elapsedLabel.text = formatElapsed(elapsedSeconds);
      } else {
        elapsedLabel.text = formatElapsed(finalElapsed);
      }
    } else {
      elapsedLabel.visibility = 'collapse' as any;
    }

    // Timer management
    if (state === 'generating' && intervalId === null) {
      startInterval();
    }
    if (state === 'complete' || state === 'error') {
      finalElapsed = elapsedSeconds;
      stopInterval();
    }
    if (state === 'idle') {
      elapsedSeconds = 0;
      finalElapsed = 0;
      stopInterval();
    }

    // Retry button
    retryBtn.visibility = (state === 'error' ? 'visible' : 'collapse') as any;

    // Accessibility
    root.automationText = `Generation ${state}`;
  }

  // Initial sync
  syncStatus();

  disposers.push(() => stopInterval());

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createGenerationIndicator;
